### PR TITLE
 Add `assert_regex`

### PR DESCRIPTION
**Pull request by rico-chet**
Add a function wrapping the commonly used `[[ value =~ pattern ]]`
construct, printing both operands in case of failure.

This addresses the need in #7 without affecting backward compatibility
of `assert_equal`.

### CODE REVIEW

**Comment by martin-schulze-vireso** [comment]
I am unsure about potential code sharing with `assert_output --regex`.

**Comment by martin-schulze-vireso** [comment]
Pleased add a test for the invalid regex branch.

File: src/assert_regex.bash [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#
+# If the value is longer than one line then it is displayed in *multi-line*
+# format.
+assert_regex() {
+	local -r value="${1}"
+	local -r pattern="${2}"
+
+	if [[ '' =~ ${pattern} ]] || (( ${?} == 2 )); then
+		echo "Invalid extended regular expression: \`${pattern}'" \
+		| batslib_decorate 'ERROR: assert_output' \
```
Comment (martin-schulze-vireso) [inline src/assert_regex.bash:32]: This should be assert_regex?

**Review by martin-schulze-vireso — CHANGES_REQUESTED** [review]

**Comment by rico-chet** [comment]
I added the error case test and fixed a copy-paste issue.

I also found that the multi-line value case needs more test coverage and documentation. It's not obvious that the assert succeeds when _any_ line of the `value` matches the `pattern`.

**Comment by rico-chet** [comment]
> > It's not obvious that the assert succeeds when _any_ line of the `value` matches the `pattern`.
> 
> I would have thought that it is default regex behavior to Match anywhere in the string If there are no begin/end markers ($^).

This is the case, but I don't know by heart how multi-line stuff is handled exactly, so I added a couple characterization tests.

**Comment by rico-chet** [comment]
> I am unsure about potential code sharing with `assert_output --regex`. Then again its not even 20 lines...

`assert_output` only acts on the `$output` content, but not on e.g. `$stderr` or any other data like a call trace from a mock script (the scenario where I used this last time). This is the rationale for not re-using `assert_output` for this. Also, I don't like all the parameters as they clutter tests in the long run.

**Comment by rico-chet** [comment]
should be good to go now

**Review by martin-schulze-vireso — APPROVED** [review]

**Comment by martin-schulze-vireso** [comment]
Should we also provide refute_regex in this PR?

File: README.md [hunk 2]
```diff
@@ -693,6 +693,62 @@ An error is displayed if the specified extended regular expression is invalid.
...
+-- values does not match regular expression --
+value    : what
+pattern  : x$
+--
+```
+
+If the value is longer than one line then it is displayed in *multi-line*
+format.
+
+Multi-line value matches a single-line pattern when one of its lines matches,
```
[the remark this answers is not in the document]
Comment (rico-chet) [reply README.md:722]: I get what you mean (`\n` in the value matches `.` in the pattern) and I was confused as I don't use multi-line matching often. The message needs an improvement, I agree. I will see how fast I can do it.

**Comment by rico-chet** [comment]
> Should we also provide refute_regex in this PR?

I prefer a separate PR, to keep the batch size small and get things done faster.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rico-chet) [reply README.md:722]: I am planning to simply refer to the Bash docs: https://www.gnu.org/software/bash/manual/html_node/Conditional-Constructs.html#:~:text=[[…]], as it's rather complicated. The applied case sensitivity should also be made visible to users, IMHO. `BASH_REMATCH` visibility should be part of a test, too.

**Review by martin-schulze-vireso — APPROVED** [review]

**Comment by martin-schulze-vireso** [comment]
Thanks for your contribution.

**Pull request merged**